### PR TITLE
[ALLUXIO-1896] test equals and hashCode for options

### PR DESCRIPTION
**Pull request by aaudiber**
https://alluxio.atlassian.net/browse/ALLUXIO-1896

### CODE REVIEW

**Comment by ifcharming** [comment]
Thanks for adding the tests! Can you resolve the test failures? Otherwise LGTM

File: core/common/src/test/java/alluxio/CommonTestUtils.java [hunk 1]
```diff
@@ -54,4 +78,62 @@
...
+   * It is required that the given class has a no-arg constructor.
+   *
+   * Note: To use this method to test a class which contains a final class as a field, you must
+   * prepare the final class for testing. See the top of {@link CommonTestUtilsTest} for an example.
+   *
+   * @param clazz the class to test the equals and hashCode methods for
+   * @param excludedFields names of fields which should not impact equality
+   */
+  public static <T> void testEquals(Class<T> clazz, String... excludedFields) throws Exception {
+    Set<String> excludedFieldsSet = Sets.newHashSet(Arrays.asList(excludedFields));
```
Comment (yupeng9) [inline core/common/src/test/java/alluxio/CommonTestUtils.java]: I wonder whats the limitation of this test. Would it support the fields of references that lead to a cyclic graph?

**Comment by yupeng9** [comment]
This is cool! I left one question.

File: core/client/src/test/java/alluxio/client/file/options/CreateDirectoryOptionsTest.java [hunk 2]
```diff
@@ -11,19 +11,26 @@
...
+import org.powermock.core.classloader.annotations.PrepareForTest;
+import org.powermock.modules.junit4.PowerMockRunner;
 
 import java.util.Random;
 
 /**
  * Tests for the {@link CreateDirectoryOptions} class.
  */
+@RunWith(PowerMockRunner.class)
+@PrepareForTest(UnderStorageType.class)
```
Comment (jsimsa) [inline core/client/src/test/java/alluxio/client/file/options/CreateDirectoryOptionsTest.java]: would using CreateDirectoryOptions.class here instead work? if not, I would add a comment here to explain why we need to do this ... similarly for other tests in this PR

**Comment by jsimsa** [comment]
@aaudiber thank you for this PR, left one comment ... could we give the lineage options similar treatment (toString, equals, hashCode, and tests)?

File: core/common/src/test/java/alluxio/CommonTestUtils.java (hunk 1 above)
Comment (aaudiber) [inline core/common/src/test/java/alluxio/CommonTestUtils.java]: This test doesn't touch fields of fields, so there's no danger of hitting a cycle. The two values used are `null` and a powermock instance - no real value is instantiated.

**Comment by aaudiber** [comment]
@jsimsa sgtm, I'd rather do it in another commit though since this one is already pretty big.

File: core/server/src/test/java/alluxio/master/file/options/CreateDirectoryOptionsTest.java [hunk 3]
```diff
@@ -29,6 +29,7 @@
  * Unit tests for {@link CreateDirectoryOptions}.
  */
 @RunWith(PowerMockRunner.class)
+// Need to mock PermissionStatus to use CommonTestUtils#testEquals.
 @PrepareForTest(PermissionStatus.class)
```
Comment (jsimsa) [inline core/server/src/test/java/alluxio/master/file/options/CreateDirectoryOptionsTest.java]: why is this not needed by other test classes that use testEquals?

File: core/server/src/test/java/alluxio/master/file/options/CreateFileOptionsTest.java [hunk 4]
```diff
@@ -29,6 +29,7 @@
  * Unit tests for {@link CreateFileOptions}.
  */
 @RunWith(PowerMockRunner.class)
+//Need to mock PermissionStatus to use CommonTestUtils#testEquals.
```
Comment (jsimsa) [inline core/server/src/test/java/alluxio/master/file/options/CreateFileOptionsTest.java]: ditto

File: core/server/src/test/java/alluxio/master/file/options/CreateDirectoryOptionsTest.java (hunk 3 above)
Comment (aaudiber) [inline core/server/src/test/java/alluxio/master/file/options/CreateDirectoryOptionsTest.java]: the javadoc for testEquals mentions this - basically `@PrepareForTest` is used as a last resort if (1) the field isn't a primitive type (2) the field isn't an enum type (3) the field type is final (4) the field type doesn't have a no-arg constructor. Unfortunately, these happen to all be true for `PermissionStatus`.

Comment (jsimsa) [inline core/server/src/test/java/alluxio/master/file/options/CreateDirectoryOptionsTest.java]: sgtm

File: core/server/src/test/java/alluxio/master/file/options/CreateFileOptionsTest.java (hunk 4 above)
Comment (jsimsa) [inline core/server/src/test/java/alluxio/master/file/options/CreateFileOptionsTest.java]: nit: add a space after `//`

**Comment by jsimsa** [comment]
just one minor nit, looks good to me otherwise

**Comment by aaudiber** [comment]
@jsimsa done

**Comment by jsimsa** [comment]
LGTM

**Pull request merged**